File: 滑动窗口/lengthOfLongestSubstring.js
// 求一个字符串s中不含有重复字符的最长子串的长度
function lengthOfLongestSubstring(s) {
    if (s.length === 0) return 0
    const map = new Map()
    let max = 0, left = 0
    // i为右边界
    for (let i = 0; i < s.length; i++) {
        // 修改子串的起始位置
        if (map.has(s.charAt(i))) left = Math.max(left, map.get(s.charAt(i))+1)
        // 更新下标
        map.set(s.charAt(i), i)
        max = Math.max(max, i-left+1)
    }
    return max
}

/**
 * 思路
 * 使用map记录字符的下标
 */

File: slidingWindow/lengthOfLongestSubstring.go
package main

// 求一个字符串s中不含有重复字符的最长子串的长度
func lengthOfLongestSubstring(s string) int {
	m := map[byte]int{}
	n := len(s)
	if n == 0 {
		return 0
	}
	var max = 0
	var left = 0
	for i := 0; i < len(s); i++ {
		if v, ok := m[s[i]]; ok {
			left = Max(left, v+1)
		}
		m[s[i]] = i
		max = Max(max, i-left+1)
	}
	return max
}

func Max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

// func main() {
// 	s := "abba"
// 	ret := lengthOfLongestSubstring(s)
// 	fmt.Printf("%v", ret)
// }
